# Text classification

Text classification is a common NLP task that assigns a label or class to text. Some of the largest companies run text classification in production for a wide range of practical applications. One of the most popular forms of text classification is sentiment analysis, which assigns a label like 🙂 positive, 🙁 negative, or 😐 neutral to a sequence of text.

This guide will show you how to:

1. Finetune [DistilBERT](https://huggingface.co/distilbert-base-uncased) on the [IMDb](https://huggingface.co/datasets/imdb) dataset to determine whether a movie review is positive or negative.
2. Use your finetuned model for inference.

<Tip>
The task illustrated in this tutorial is supported by the following model architectures:

<!--This tip is automatically generated by `make fix-copies`, do not fill manually!-->

[ALBERT](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/albert), [BART](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/bart), [BERT](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/bert), [BigBird](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/big_bird), [BigBird-Pegasus](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/bigbird_pegasus), [BioGpt](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/biogpt), [BLOOM](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/bloom), [CamemBERT](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/camembert), [CANINE](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/canine), [ConvBERT](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/convbert), [CTRL](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/ctrl), [Data2VecText](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/data2vec-text), [DeBERTa](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/deberta), [DeBERTa-v2](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/deberta-v2), [DistilBERT](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/distilbert), [ELECTRA](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/electra), [ERNIE](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/ernie), [ErnieM](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/ernie_m), [ESM](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/esm), [FlauBERT](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/flaubert), [FNet](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/fnet), [Funnel Transformer](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/funnel), [GPT-Sw3](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/gpt-sw3), [OpenAI GPT-2](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/gpt2), [GPTBigCode](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/gpt_bigcode), [GPT Neo](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/gpt_neo), [GPT NeoX](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/gpt_neox), [GPT-J](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/gptj), [I-BERT](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/ibert), [LayoutLM](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/layoutlm), [LayoutLMv2](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/layoutlmv2), [LayoutLMv3](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/layoutlmv3), [LED](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/led), [LiLT](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/lilt), [LLaMA](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/llama), [Longformer](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/longformer), [LUKE](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/luke), [MarkupLM](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/markuplm), [mBART](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/mbart), [MEGA](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/mega), [Megatron-BERT](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/megatron-bert), [MobileBERT](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/mobilebert), [MPNet](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/mpnet), [MVP](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/mvp), [Nezha](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/nezha), [Nyströmformer](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/nystromformer), [OpenLlama](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/open-llama), [OpenAI GPT](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/openai-gpt), [OPT](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/opt), [Perceiver](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/perceiver), [PLBart](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/plbart), [QDQBert](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/qdqbert), [Reformer](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/reformer), [RemBERT](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/rembert), [RoBERTa](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/roberta), [RoBERTa-PreLayerNorm](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/roberta-prelayernorm), [RoCBert](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/roc_bert), [RoFormer](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/roformer), [SqueezeBERT](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/squeezebert), [TAPAS](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/tapas), [Transformer-XL](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/transfo-xl), [XLM](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/xlm), [XLM-RoBERTa](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/xlm-roberta), [XLM-RoBERTa-XL](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/xlm-roberta-xl), [XLNet](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/xlnet), [X-MOD](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/xmod), [YOSO](https://huggingface.co/docs/transformers/main/en/tasks/../model_doc/yoso)


<!--End of the generated tip-->

</Tip>

Before you begin, make sure you have all the necessary libraries installed:

```bash
pip install transformers datasets evaluate
```

We encourage you to login to your Hugging Face account so you can upload and share your model with the community. When prompted, enter your token to login:

In [1]:
from huggingface_hub import notebook_login

notebook_login()

#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

## Load IMDb dataset

Start by loading the IMDb dataset from the 🤗 Datasets library:

In [2]:
from datasets import load_dataset
imdb = load_dataset("imdb")

Then take a look at an example:

In [3]:
imdb["test"][0]

{'text': 'I love sci-fi and am willing to put up with a lot. Sci-fi movies/TV are usually underfunded, under-appreciated and misunderstood. I tried to like this, I really did, but it is to good TV sci-fi as Babylon 5 is to Star Trek (the original). Silly prosthetics, cheap cardboard sets, stilted dialogues, CG that doesn\'t match the background, and painfully one-dimensional characters cannot be overcome with a \'sci-fi\' setting. (I\'m sure there are those of you out there who think Babylon 5 is good sci-fi TV. It\'s not. It\'s clichéd and uninspiring.) While US viewers might like emotion and character development, sci-fi is a genre that does not take itself seriously (cf. Star Trek). It may treat important issues, yet not as a serious philosophy. It\'s really difficult to care about the characters here as they are not simply foolish, just missing a spark of life. Their actions and reactions are wooden and predictable, often painful to watch. The makers of Earth KNOW it\'s rubbish as 

There are two fields in this dataset:

- `text`: the movie review text.
- `label`: a value that is either `0` for a negative review or `1` for a positive review.

## Preprocess

The next step is to load a DistilBERT tokenizer to preprocess the `text` field:

In [4]:
from transformers import AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("distilbert-base-uncased")

Create a preprocessing function to tokenize `text` and truncate sequences to be no longer than DistilBERT's maximum input length:

In [5]:
def preprocess_function(examples):
    return tokenizer(examples["text"], truncation=True)

# Add different paddings

In [6]:
from datasets import load_dataset
imdb = load_dataset("imdb")

n_range = range(0, 101, 10)

def add_text(example, n):
  # example is a dictionary with keys 'text' and 'label'
  # we can modify the 'text' value and return the modified example
  example['text'] = "[PAD]" * n + example['text']
  return example

for n in n_range:
    # use format method to create a variable name with n
    var_name = "imdb_padding_{}".format(n)
    # use exec function to execute the assignment statement
    exec("{} = imdb.map(lambda x: add_text(x, n))".format(var_name))
    print(var_name, len(eval(var_name)))

imdb_padding_0 3
imdb_padding_10 3
imdb_padding_20 3
imdb_padding_30 3
imdb_padding_40 3
imdb_padding_50 3
imdb_padding_60 3
imdb_padding_70 3
imdb_padding_80 3
imdb_padding_90 3
imdb_padding_100 3


In [7]:
imdb_padding_10["test"][0]

{'text': '[PAD][PAD][PAD][PAD][PAD][PAD][PAD][PAD][PAD][PAD]I love sci-fi and am willing to put up with a lot. Sci-fi movies/TV are usually underfunded, under-appreciated and misunderstood. I tried to like this, I really did, but it is to good TV sci-fi as Babylon 5 is to Star Trek (the original). Silly prosthetics, cheap cardboard sets, stilted dialogues, CG that doesn\'t match the background, and painfully one-dimensional characters cannot be overcome with a \'sci-fi\' setting. (I\'m sure there are those of you out there who think Babylon 5 is good sci-fi TV. It\'s not. It\'s clichéd and uninspiring.) While US viewers might like emotion and character development, sci-fi is a genre that does not take itself seriously (cf. Star Trek). It may treat important issues, yet not as a serious philosophy. It\'s really difficult to care about the characters here as they are not simply foolish, just missing a spark of life. Their actions and reactions are wooden and predictable, often painful to

To apply the preprocessing function over the entire dataset, use 🤗 Datasets [map](https://huggingface.co/docs/datasets/main/en/package_reference/main_classes#datasets.Dataset.map) function. You can speed up `map` by setting `batched=True` to process multiple elements of the dataset at once:

In [8]:
from transformers import AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("distilbert-base-uncased")

def preprocess_function(examples):
    return tokenizer(examples["text"], truncation=True)

In [9]:
for n in n_range:
    filename = f"imdb_padding_{n}"
    file = eval(filename)
    tokenized_filename = f"tokenized_imdb_{n}"
    exec(f"{tokenized_filename} = file.map(preprocess_function, batched=True)")
    tokenized_file = eval(tokenized_filename)

Map:   0%|          | 0/25000 [00:00<?, ? examples/s]

Now create a batch of examples using [DataCollatorWithPadding](https://huggingface.co/docs/transformers/main/en/main_classes/data_collator#transformers.DataCollatorWithPadding). It's more efficient to *dynamically pad* the sentences to the longest length in a batch during collation, instead of padding the whole dataset to the maximum length.

In [10]:
from transformers import DataCollatorWithPadding
data_collator = DataCollatorWithPadding(tokenizer=tokenizer)

## Evaluate

Including a metric during training is often helpful for evaluating your model's performance. You can quickly load a evaluation method with the 🤗 [Evaluate](https://huggingface.co/docs/evaluate/index) library. For this task, load the [accuracy](https://huggingface.co/spaces/evaluate-metric/accuracy) metric (see the 🤗 Evaluate [quick tour](https://huggingface.co/docs/evaluate/a_quick_tour) to learn more about how to load and compute a metric):

In [11]:
import evaluate
accuracy = evaluate.load("accuracy")

Then create a function that passes your predictions and labels to [compute](https://huggingface.co/docs/evaluate/main/en/package_reference/main_classes#evaluate.EvaluationModule.compute) to calculate the accuracy:

In [12]:
import numpy as np
def compute_metrics(eval_pred):
    predictions, labels = eval_pred
    predictions = np.argmax(predictions, axis=1)
    return accuracy.compute(predictions=predictions, references=labels)

Your `compute_metrics` function is ready to go now, and you'll return to it when you setup your training.

## Train

Before you start training your model, create a map of the expected ids to their labels with `id2label` and `label2id`:

In [13]:
id2label = {0: "NEGATIVE", 1: "POSITIVE"}
label2id = {"NEGATIVE": 0, "POSITIVE": 1}

<Tip>

If you aren't familiar with finetuning a model with the [Trainer](https://huggingface.co/docs/transformers/main/en/main_classes/trainer#transformers.Trainer), take a look at the basic tutorial [here](https://huggingface.co/docs/transformers/main/en/tasks/../training#train-with-pytorch-trainer)!

</Tip>

You're ready to start training your model now! Load DistilBERT with [AutoModelForSequenceClassification](https://huggingface.co/docs/transformers/main/en/model_doc/auto#transformers.AutoModelForSequenceClassification) along with the number of expected labels, and the label mappings:

In [14]:
from transformers import AutoModelForSequenceClassification, TrainingArguments, Trainer

model = AutoModelForSequenceClassification.from_pretrained(
    "distilbert-base-uncased", num_labels=2, id2label=id2label, label2id=label2id
)

Some weights of DistilBertForSequenceClassification were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['classifier.weight', 'pre_classifier.weight', 'classifier.bias', 'pre_classifier.bias']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


# Train() with different paddings

In [15]:
import json
results = {}

In [16]:
# define the range of n
#for n in n_range:
for n in n_range:
    # use format method to create a variable name with n
    variable = f"tokenized_imdb_{n}"
    output_dir = f"left_padding{n}model"
    logging_dir=f"left_padding{n}model_logs"

    # use exec function to execute the assignment statement
    var = eval(variable)
    training_args = TrainingArguments(
        
        learning_rate=2e-5,
        per_device_train_batch_size=16,
        per_device_eval_batch_size=16,
        num_train_epochs=20,
        weight_decay=0.01,
        
        evaluation_strategy="epoch",
        save_strategy="epoch",

        push_to_hub=True,
        seed=42,
        data_seed=123,
        output_dir=output_dir,
        logging_dir="./logs",
        logging_steps=100
    )
    
    # define the trainer
    trainer = Trainer(
        model=model,
        args=training_args,
        train_dataset=var["train"],
        eval_dataset=var["test"],
        tokenizer=tokenizer,
        data_collator=data_collator,
        compute_metrics=compute_metrics,
    )

    trainer.train(resume_from_checkpoint=True)
    trainer.push_to_hub()
    
    result = trainer.evaluate()
    results[n] = result



  0%|          | 0/156300 [00:00<?, ?it/s]

You're using a DistilBertTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.


{'loss': 0.0036, 'learning_rate': 1.599488163787588e-05, 'epoch': 20.03}
{'loss': 0.0141, 'learning_rate': 1.598208573256558e-05, 'epoch': 20.09}
{'loss': 0.0001, 'learning_rate': 1.596928982725528e-05, 'epoch': 20.15}
{'loss': 0.0062, 'learning_rate': 1.595649392194498e-05, 'epoch': 20.22}
{'loss': 0.0116, 'learning_rate': 1.594369801663468e-05, 'epoch': 20.28}
{'loss': 0.0165, 'learning_rate': 1.5930902111324377e-05, 'epoch': 20.35}
{'loss': 0.0102, 'learning_rate': 1.5918106206014075e-05, 'epoch': 20.41}
{'loss': 0.0011, 'learning_rate': 1.5905310300703776e-05, 'epoch': 20.47}
{'loss': 0.0045, 'learning_rate': 1.5892514395393474e-05, 'epoch': 20.54}
{'loss': 0.0013, 'learning_rate': 1.5879718490083176e-05, 'epoch': 20.6}
{'loss': 0.0117, 'learning_rate': 1.5866922584772874e-05, 'epoch': 20.67}
{'loss': 0.0261, 'learning_rate': 1.5854126679462575e-05, 'epoch': 20.73}
{'loss': 0.0099, 'learning_rate': 1.5841330774152273e-05, 'epoch': 20.79}
{'loss': 0.0121, 'learning_rate': 1.58285348

  0%|          | 0/1563 [00:00<?, ?it/s]

{'eval_loss': 0.7199747562408447, 'eval_accuracy': 0.92476, 'eval_runtime': 367.458, 'eval_samples_per_second': 68.035, 'eval_steps_per_second': 4.254, 'epoch': 21.0}
{'loss': 0.0108, 'learning_rate': 1.5790147152911072e-05, 'epoch': 21.05}
{'loss': 0.0219, 'learning_rate': 1.577735124760077e-05, 'epoch': 21.11}
{'loss': 0.0056, 'learning_rate': 1.5764555342290468e-05, 'epoch': 21.18}
{'loss': 0.0001, 'learning_rate': 1.575175943698017e-05, 'epoch': 21.24}
{'loss': 0.0125, 'learning_rate': 1.5738963531669867e-05, 'epoch': 21.31}
{'loss': 0.0172, 'learning_rate': 1.5726167626359565e-05, 'epoch': 21.37}
{'loss': 0.0125, 'learning_rate': 1.5713371721049266e-05, 'epoch': 21.43}
{'loss': 0.0141, 'learning_rate': 1.5700575815738964e-05, 'epoch': 21.5}
{'loss': 0.0078, 'learning_rate': 1.5687779910428666e-05, 'epoch': 21.56}
{'loss': 0.0063, 'learning_rate': 1.5674984005118364e-05, 'epoch': 21.63}
{'loss': 0.0075, 'learning_rate': 1.566218809980806e-05, 'epoch': 21.69}
{'loss': 0.0122, 'learn

  0%|          | 0/1563 [00:00<?, ?it/s]

{'eval_loss': 0.6553881168365479, 'eval_accuracy': 0.92208, 'eval_runtime': 368.8167, 'eval_samples_per_second': 67.784, 'eval_steps_per_second': 4.238, 'epoch': 22.0}
{'loss': 0.0139, 'learning_rate': 1.559820857325656e-05, 'epoch': 22.01}
{'loss': 0.0003, 'learning_rate': 1.558541266794626e-05, 'epoch': 22.07}
{'loss': 0.0, 'learning_rate': 1.5572616762635958e-05, 'epoch': 22.14}
{'loss': 0.003, 'learning_rate': 1.555982085732566e-05, 'epoch': 22.2}
{'loss': 0.0095, 'learning_rate': 1.5547024952015357e-05, 'epoch': 22.26}
{'loss': 0.0035, 'learning_rate': 1.5534229046705055e-05, 'epoch': 22.33}
{'loss': 0.0066, 'learning_rate': 1.5521433141394756e-05, 'epoch': 22.39}
{'loss': 0.0, 'learning_rate': 1.5508637236084454e-05, 'epoch': 22.46}
{'loss': 0.0059, 'learning_rate': 1.5495841330774152e-05, 'epoch': 22.52}
{'loss': 0.0148, 'learning_rate': 1.548304542546385e-05, 'epoch': 22.58}
{'loss': 0.0031, 'learning_rate': 1.547024952015355e-05, 'epoch': 22.65}
{'loss': 0.0, 'learning_rate': 

  0%|          | 0/1563 [00:00<?, ?it/s]

{'eval_loss': 0.6918370127677917, 'eval_accuracy': 0.92484, 'eval_runtime': 368.414, 'eval_samples_per_second': 67.858, 'eval_steps_per_second': 4.243, 'epoch': 23.0}
{'loss': 0.0094, 'learning_rate': 1.539347408829175e-05, 'epoch': 23.03}
{'loss': 0.0159, 'learning_rate': 1.5380678182981448e-05, 'epoch': 23.1}
{'loss': 0.0032, 'learning_rate': 1.5367882277671146e-05, 'epoch': 23.16}
{'loss': 0.0107, 'learning_rate': 1.5355086372360847e-05, 'epoch': 23.22}
{'loss': 0.0078, 'learning_rate': 1.5342290467050545e-05, 'epoch': 23.29}
{'loss': 0.0, 'learning_rate': 1.5329494561740243e-05, 'epoch': 23.35}
{'loss': 0.0058, 'learning_rate': 1.5316698656429944e-05, 'epoch': 23.42}
{'loss': 0.009, 'learning_rate': 1.5303902751119642e-05, 'epoch': 23.48}
{'loss': 0.0025, 'learning_rate': 1.5291106845809344e-05, 'epoch': 23.54}
{'loss': 0.0064, 'learning_rate': 1.527831094049904e-05, 'epoch': 23.61}
{'loss': 0.0057, 'learning_rate': 1.526551503518874e-05, 'epoch': 23.67}
{'loss': 0.011, 'learning_r

  0%|          | 0/1563 [00:00<?, ?it/s]

{'eval_loss': 0.7658491134643555, 'eval_accuracy': 0.9264, 'eval_runtime': 367.4719, 'eval_samples_per_second': 68.032, 'eval_steps_per_second': 4.253, 'epoch': 24.0}


SafetensorError: Error while serializing: IoError(Os { code: 112, kind: StorageFull, message: "磁盘空间不足。" })

In [ ]:
with open("results.json", "w") as f:
    json.dump(results, f)

## padding 50, 100, 150

In [ ]:
training_args = TrainingArguments(
    output_dir="left_padding50_model",
    learning_rate=2e-5,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=16,
    num_train_epochs=2,
    weight_decay=0.01,
    evaluation_strategy="epoch",
    save_strategy="epoch",
    load_best_model_at_end=True,
    push_to_hub=True,
    seed=42,
    data_seed=123
)

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_imdb_50["train"],
    eval_dataset=tokenized_imdb_50["test"],
    tokenizer=tokenizer,
    data_collator=data_collator,
    compute_metrics=compute_metrics,
)

trainer.train()

  0%|          | 0/3126 [00:00<?, ?it/s]

You're using a DistilBertTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.


{'loss': 0.3348, 'learning_rate': 1.6801023672424827e-05, 'epoch': 0.32}
{'loss': 0.2586, 'learning_rate': 1.3602047344849649e-05, 'epoch': 0.64}
{'loss': 0.2306, 'learning_rate': 1.0403071017274472e-05, 'epoch': 0.96}


  0%|          | 0/1563 [00:00<?, ?it/s]

{'eval_loss': 0.26562219858169556, 'eval_accuracy': 0.90268, 'eval_runtime': 377.3093, 'eval_samples_per_second': 66.259, 'eval_steps_per_second': 4.142, 'epoch': 1.0}
{'loss': 0.1622, 'learning_rate': 7.204094689699297e-06, 'epoch': 1.28}
{'loss': 0.1487, 'learning_rate': 4.005118362124121e-06, 'epoch': 1.6}
{'loss': 0.1599, 'learning_rate': 8.061420345489445e-07, 'epoch': 1.92}


  0%|          | 0/1563 [00:00<?, ?it/s]

{'eval_loss': 0.235260471701622, 'eval_accuracy': 0.92952, 'eval_runtime': 376.2296, 'eval_samples_per_second': 66.449, 'eval_steps_per_second': 4.154, 'epoch': 2.0}
{'train_runtime': 2855.1642, 'train_samples_per_second': 17.512, 'train_steps_per_second': 1.095, 'train_loss': 0.2124700436436512, 'epoch': 2.0}


TrainOutput(global_step=3126, training_loss=0.2124700436436512, metrics={'train_runtime': 2855.1642, 'train_samples_per_second': 17.512, 'train_steps_per_second': 1.095, 'train_loss': 0.2124700436436512, 'epoch': 2.0})

In [ ]:
trainer.push_to_hub()

'https://huggingface.co/Realgon/left_padding50_model/tree/main/'

In [ ]:
training_args = TrainingArguments(
    output_dir="left_padding100_model",
    learning_rate=2e-5,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=16,
    num_train_epochs=2,
    weight_decay=0.01,
    evaluation_strategy="epoch",
    save_strategy="epoch",
    load_best_model_at_end=True,
    push_to_hub=True,
    seed=42,
    data_seed=123
)

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_imdb_100["train"],
    eval_dataset=tokenized_imdb_100["test"],
    tokenizer=tokenizer,
    data_collator=data_collator,
    compute_metrics=compute_metrics,

)

trainer.train()

  0%|          | 0/3126 [00:00<?, ?it/s]

{'loss': 0.1975, 'learning_rate': 1.6801023672424827e-05, 'epoch': 0.32}
{'loss': 0.1874, 'learning_rate': 1.3602047344849649e-05, 'epoch': 0.64}
{'loss': 0.1809, 'learning_rate': 1.0403071017274472e-05, 'epoch': 0.96}


  0%|          | 0/1563 [00:00<?, ?it/s]

{'eval_loss': 0.2602356970310211, 'eval_accuracy': 0.9228, 'eval_runtime': 373.1317, 'eval_samples_per_second': 67.0, 'eval_steps_per_second': 4.189, 'epoch': 1.0}
{'loss': 0.1007, 'learning_rate': 7.204094689699297e-06, 'epoch': 1.28}
{'loss': 0.0853, 'learning_rate': 4.005118362124121e-06, 'epoch': 1.6}
{'loss': 0.1099, 'learning_rate': 8.061420345489445e-07, 'epoch': 1.92}


  0%|          | 0/1563 [00:00<?, ?it/s]

{'eval_loss': 0.284171462059021, 'eval_accuracy': 0.93, 'eval_runtime': 372.6626, 'eval_samples_per_second': 67.085, 'eval_steps_per_second': 4.194, 'epoch': 2.0}
{'train_runtime': 2858.2215, 'train_samples_per_second': 17.493, 'train_steps_per_second': 1.094, 'train_loss': 0.1417671253646099, 'epoch': 2.0}


TrainOutput(global_step=3126, training_loss=0.1417671253646099, metrics={'train_runtime': 2858.2215, 'train_samples_per_second': 17.493, 'train_steps_per_second': 1.094, 'train_loss': 0.1417671253646099, 'epoch': 2.0})

In [ ]:
trainer.push_to_hub()

'https://huggingface.co/Realgon/left_padding100_model/tree/main/'

At this point, only three steps remain:

1. Define your training hyperparameters in [TrainingArguments](https://huggingface.co/docs/transformers/main/en/main_classes/trainer#transformers.TrainingArguments). The only required parameter is `output_dir` which specifies where to save your model. You'll push this model to the Hub by setting `push_to_hub=True` (you need to be signed in to Hugging Face to upload your model). At the end of each epoch, the [Trainer](https://huggingface.co/docs/transformers/main/en/main_classes/trainer#transformers.Trainer) will evaluate the accuracy and save the training checkpoint.
2. Pass the training arguments to [Trainer](https://huggingface.co/docs/transformers/main/en/main_classes/trainer#transformers.Trainer) along with the model, dataset, tokenizer, data collator, and `compute_metrics` function.
3. Call [train()](https://huggingface.co/docs/transformers/main/en/main_classes/trainer#transformers.Trainer.train) to finetune your model.

In [ ]:
training_args = TrainingArguments(
    output_dir="left_padding150_model",
    learning_rate=2e-5,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=16,
    num_train_epochs=2,
    weight_decay=0.01,

    evaluation_strategy="epoch",
    save_strategy="epoch",
    
    load_best_model_at_end=True,
    push_to_hub=True,
    seed=42,
    data_seed=123)

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_imdb_150["train"],
    eval_dataset=tokenized_imdb_150["test"],
    tokenizer=tokenizer,
    data_collator=data_collator,
    compute_metrics=compute_metrics,

)

trainer.train()

  0%|          | 0/3126 [00:00<?, ?it/s]

{'loss': 0.1597, 'learning_rate': 1.6801023672424827e-05, 'epoch': 0.32}
{'loss': 0.1511, 'learning_rate': 1.3602047344849649e-05, 'epoch': 0.64}
{'loss': 0.1564, 'learning_rate': 1.0403071017274472e-05, 'epoch': 0.96}


  0%|          | 0/1563 [00:00<?, ?it/s]

{'eval_loss': 0.2768949866294861, 'eval_accuracy': 0.92112, 'eval_runtime': 379.0743, 'eval_samples_per_second': 65.95, 'eval_steps_per_second': 4.123, 'epoch': 1.0}
{'loss': 0.082, 'learning_rate': 7.204094689699297e-06, 'epoch': 1.28}
{'loss': 0.0764, 'learning_rate': 4.005118362124121e-06, 'epoch': 1.6}
{'loss': 0.0916, 'learning_rate': 8.061420345489445e-07, 'epoch': 1.92}


  0%|          | 0/1563 [00:00<?, ?it/s]

{'eval_loss': 0.3246845006942749, 'eval_accuracy': 0.92544, 'eval_runtime': 378.4581, 'eval_samples_per_second': 66.058, 'eval_steps_per_second': 4.13, 'epoch': 2.0}
{'train_runtime': 2899.2484, 'train_samples_per_second': 17.246, 'train_steps_per_second': 1.078, 'train_loss': 0.11752973331981031, 'epoch': 2.0}


TrainOutput(global_step=3126, training_loss=0.11752973331981031, metrics={'train_runtime': 2899.2484, 'train_samples_per_second': 17.246, 'train_steps_per_second': 1.078, 'train_loss': 0.11752973331981031, 'epoch': 2.0})

无法呈现“application/vnd.jupyter.widget-view+json”的内容
{"model_id":"057a11bb382d417db9ab4c042ad62e75","version_major":2,"version_minor":0}
You're using a DistilBertTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.
{'loss': 0.3392, 'learning_rate': 1.6801023672424827e-05, 'epoch': 0.32}
{'loss': 0.2638, 'learning_rate': 1.3602047344849649e-05, 'epoch': 0.64}
{'loss': 0.2446, 'learning_rate': 1.0403071017274472e-05, 'epoch': 0.96}
无法呈现“application/vnd.jupyter.widget-view+json”的内容
{"model_id":"ea2324d6b1874a199e5f632294202495","version_major":2,"version_minor":0}
{'eval_loss': 0.3612774908542633, 'eval_accuracy': 0.86776, 'eval_runtime': 1436.2269, 'eval_samples_per_second': 17.407, 'eval_steps_per_second': 1.088, 'epoch': 1.0}
{'loss': 0.1764, 'learning_rate': 7.204094689699297e-06, 'epoch': 1.28}
{'loss': 0.1591, 'learning_rate': 4.005118362124121e-06, 'epoch': 1.6}
{'loss': 0.1746, 'learning_rate': 8.061420345489445e-07, 'epoch': 1.92}
无法呈现“application/vnd.jupyter.widget-view+json”的内容
{"model_id":"e11f558dc7c64eacb2d28408bf424940","version_major":2,"version_minor":0}
{'eval_loss': 0.24108850955963135, 'eval_accuracy': 0.92488, 'eval_runtime': 746.9998, 'eval_samples_per_second': 33.467, 'eval_steps_per_second': 2.092, 'epoch': 2.0}
{'train_runtime': 6886.1501, 'train_samples_per_second': 7.261, 'train_steps_per_second': 0.454, 'train_loss': 0.22320297629270353, 'epoch': 2.0}
TrainOutput(global_step=3126, training_loss=0.22320297629270353, metrics={'train_runtime': 6886.1501, 'train_samples_per_second': 7.261, 'train_steps_per_second': 0.454, 'train_loss': 0.22320297629270353, 'epoch': 2.0})

In [ ]:
trainer.push_to_hub()

'https://huggingface.co/Realgon/left_padding150_model/tree/main/'

<Tip>

[Trainer](https://huggingface.co/docs/transformers/main/en/main_classes/trainer#transformers.Trainer) applies dynamic padding by default when you pass `tokenizer` to it. In this case, you don't need to specify a data collator explicitly.

</Tip>

Once training is completed, share your model to the Hub with the [push_to_hub()](https://huggingface.co/docs/transformers/main/en/main_classes/trainer#transformers.Trainer.push_to_hub) method so everyone can use your model:

<Tip>

For a more in-depth example of how to finetune a model for text classification, take a look at the corresponding
[PyTorch notebook](https://colab.research.google.com/github/huggingface/notebooks/blob/main/examples/text_classification.ipynb)
or [TensorFlow notebook](https://colab.research.google.com/github/huggingface/notebooks/blob/main/examples/text_classification-tf.ipynb).

</Tip>

## Inference

Great, now that you've finetuned a model, you can use it for inference!

Grab some text you'd like to run inference on:

In [ ]:
text = "This was a masterpiece. Not completely faithful to the books, but enthralling from beginning to end. Might be my favorite of the three."

The simplest way to try out your finetuned model for inference is to use it in a [pipeline()](https://huggingface.co/docs/transformers/main/en/main_classes/pipelines#transformers.pipeline). Instantiate a `pipeline` for sentiment analysis with your model, and pass your text to it:

In [ ]:
from transformers import pipeline

classifier = pipeline("sentiment-analysis", model="Realgon/left_padding50_model")
classifier(text)

c:\Users\wyzhw\anaconda3\envs\pytorch2\lib\site-packages\huggingface_hub\file_download.py:137: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\wyzhw\.cache\huggingface\hub. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to see activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)


[{'label': 'POSITIVE', 'score': 0.9963740706443787}]

You can also manually replicate the results of the `pipeline` if you'd like:

Tokenize the text and return PyTorch tensors:

In [ ]:
from transformers import AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("Realgon/left_padding50_model")
inputs = tokenizer(text, return_tensors="pt")

Pass your inputs to the model and return the `logits`:

In [ ]:
from transformers import AutoModelForSequenceClassification
import torch

model = AutoModelForSequenceClassification.from_pretrained("Realgon/left_padding50_model")
with torch.no_grad():
    logits = model(**inputs).logits

Get the class with the highest probability, and use the model's `id2label` mapping to convert it to a text label:

In [ ]:
predicted_class_id = logits.argmax().item()
model.config.id2label[predicted_class_id]

'POSITIVE'